File: examples/python/mission_control/services_auth.py
"""Mission Control services with role-based access control (Chapter 5).

Same services as services.py but with requires= on each method.
Used by server.py --auth.
"""

import asyncio
from collections.abc import AsyncIterator

from aster import any_of, bidi_stream, client_stream, rpc, server_stream, service

from .roles import Role
from .types import (
    Assignment,
    Command,
    CommandResult,
    Heartbeat,
    IngestResult,
    LogEntry,
    MetricPoint,
    StatusRequest,
    StatusResponse,
    SubmitLogResult,
    TailRequest,
)

_LOG_LEVEL_RANK = {"debug": 0, "info": 1, "warn": 2, "error": 3}


@service(name="MissionControl", version=1)
class MissionControl:
    """Shared service with capability-based auth."""

    def __init__(self) -> None:
        self._log_queue: asyncio.Queue[LogEntry] = asyncio.Queue()
        self._metrics: list[MetricPoint] = []

    @rpc(requires=Role.STATUS)
    async def getStatus(self, req: StatusRequest) -> StatusResponse:
        return StatusResponse(
            agent_id=req.agent_id,
            status="running",
            uptime_secs=3600,
        )

    @rpc()
    async def submitLog(self, entry: LogEntry) -> SubmitLogResult:
        await self._log_queue.put(entry)
        return SubmitLogResult(accepted=True)

    @server_stream(requires=any_of(Role.LOGS, Role.ADMIN))
    async def tailLogs(self, req: TailRequest) -> AsyncIterator[LogEntry]:
        min_rank = _LOG_LEVEL_RANK.get(req.level.lower(), 0)
        while True:
            entry = await self._log_queue.get()
            if req.agent_id and entry.agent_id != req.agent_id:
                continue
            if _LOG_LEVEL_RANK.get(entry.level.lower(), 0) < min_rank:
                continue
            yield entry

    @client_stream(requires=Role.INGEST)
    async def ingestMetrics(
        self, stream: AsyncIterator[MetricPoint]
    ) -> IngestResult:
        accepted = 0
        async for point in stream:
            self._metrics.append(point)
            accepted += 1
        return IngestResult(accepted=accepted)


@service(name="AgentSession", version=1, scoped="session")
class AgentSession:
    """Session-scoped service with capability-based auth."""

    def __init__(self, peer: str | None = None) -> None:
        self._peer = peer
        self._agent_id = ""
        self._capabilities: list[str] = []

    @rpc(requires=Role.INGEST)
    async def register(self, hb: Heartbeat) -> Assignment:
        self._agent_id = hb.agent_id
        self._capabilities = list(hb.capabilities)
        if "gpu" in hb.capabilities:
            return Assignment(task_id="train-42", command="python train.py")
        return Assignment(task_id="idle", command="sleep 60")

    @rpc()
    async def heartbeat(self, hb: Heartbeat) -> Assignment:
        self._capabilities = list(hb.capabilities)
        return Assignment(task_id="continue", command="")

    @bidi_stream(requires=Role.ADMIN)
    async def runCommand(
        self, commands: AsyncIterator[Command]
    ) -> AsyncIterator[CommandResult]:
        async for cmd in commands:
            proc = await asyncio.create_subprocess_shell(
                cmd.command,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            stdout, stderr = await proc.communicate()
            yield CommandResult(
                stdout=stdout.decode(),
                stderr=stderr.decode(),
                exit_code=proc.returncode or 0,
            )

File: examples/typescript/missionControl/services-auth.ts
/**
 * Mission Control services with role-based access control (Chapter 5).
 *
 * Same services as services.ts but with requires= on each method.
 * Used by: bun run server.ts --auth
 */

import {
  Service,
  Rpc,
  ServerStream,
  ClientStream,
  BidiStream,
  anyOf,
} from '@aster-rpc/aster';

import {
  StatusRequest,
  StatusResponse,
  LogEntry,
  SubmitLogResult,
  TailRequest,
  MetricPoint,
  IngestResult,
  Heartbeat,
  Assignment,
  Command,
  CommandResult,
} from './types.js';

import { Role } from './roles.js';

const LOG_LEVEL_RANK: Record<string, number> = {
  debug: 0, info: 1, warn: 2, error: 3,
};

@Service({ name: "MissionControl", version: 1 })
export class MissionControl {
  private logQueue: LogEntry[] = [];
  private logWaiters: ((entry: LogEntry) => void)[] = [];
  private metrics: MetricPoint[] = [];

  @Rpc({ requires: Role.STATUS })
  async getStatus(req: StatusRequest): Promise<StatusResponse> {
    return new StatusResponse({
      agent_id: req.agent_id,
      status: "running",
      uptime_secs: 3600,
    });
  }

  @Rpc()
  async submitLog(entry: LogEntry): Promise<SubmitLogResult> {
    if (this.logWaiters.length > 0) {
      const waiter = this.logWaiters.shift()!;
      waiter(entry);
    } else {
      this.logQueue.push(entry);
    }
    return new SubmitLogResult({ accepted: true });
  }

  @ServerStream({ requires: anyOf(Role.LOGS, Role.ADMIN) })
  async *tailLogs(req: TailRequest): AsyncGenerator<LogEntry> {
    const minRank = LOG_LEVEL_RANK[req.level?.toLowerCase() ?? "info"] ?? 0;
    while (true) {
      const entry = await this.nextLog();
      if (req.agent_id && entry.agent_id !== req.agent_id) continue;
      if ((LOG_LEVEL_RANK[entry.level?.toLowerCase() ?? "info"] ?? 0) < minRank) continue;
      yield entry;
    }
  }

  private nextLog(): Promise<LogEntry> {
    const queued = this.logQueue.shift();
    if (queued) return Promise.resolve(queued);
    return new Promise<LogEntry>((resolve) => {
      this.logWaiters.push(resolve);
    });
  }

  @ClientStream({ requires: Role.INGEST })
  async ingestMetrics(stream: AsyncIterable<MetricPoint>): Promise<IngestResult> {
    let accepted = 0;
    for await (const point of stream) {
      this.metrics.push(point);
      accepted++;
    }
    return new IngestResult({ accepted });
  }
}

@Service({ name: "AgentSession", version: 1, scoped: "session" })
export class AgentSession {
  private _agentId = "";
  private _capabilities: string[] = [];

  @Rpc({ requires: Role.INGEST })
  async register(hb: Heartbeat): Promise<Assignment> {
    this._agentId = hb.agent_id;
    this._capabilities = [...(hb.capabilities ?? [])];
    if (this._capabilities.includes("gpu")) {
      return new Assignment({ task_id: "train-42", command: "python train.py" });
    }
    return new Assignment({ task_id: "idle", command: "sleep 60" });
  }

  @Rpc()
  async heartbeat(hb: Heartbeat): Promise<Assignment> {
    this._capabilities = [...(hb.capabilities ?? [])];
    return new Assignment({ task_id: "continue", command: "" });
  }

  @BidiStream({ requires: Role.ADMIN })
  async *runCommand(commands: AsyncIterable<Command>): AsyncGenerator<CommandResult> {
    for await (const cmd of commands) {
      const proc = Bun.spawn(["sh", "-c", cmd.command], {
        stdout: "pipe",
        stderr: "pipe",
      });
      const stdout = await new Response(proc.stdout).text();
      const stderr = await new Response(proc.stderr).text();
      const exit_code = await proc.exited;
      yield new CommandResult({ stdout, stderr, exit_code });
    }
  }
}
